util/chunk/iterator.go

```diff
@@ -13,40 +13,51 @@
 
 package chunk
 
+import (
+	"context"
+	"github.com/pingcap/tidb/types"
+)
+
 var (
 	_ Iterator = (*Iterator4Chunk)(nil)
 	_ Iterator = (*iterator4RowPtr)(nil)
 	_ Iterator = (*iterator4List)(nil)
 	_ Iterator = (*iterator4Slice)(nil)
 )
 
+// NoneRow indicts none row.
+var NoneRow Row
+
+// Iterable is used to iterate rows inside the object.
+type Iterable interface {
+	// Iterator returns an iterator over elements.
+	Iterator() Iterator
+}
+
 // Iterator is used to iterate a number of rows.
-//
-// for row := it.Begin(); row != it.End(); row = it.Next() {
-//     ...
-// }
 type Iterator interface {
-	// Begin resets the cursor of the iterator and returns the first Row.
-	Begin() Row
 
 	// Next returns the next Row.
 	Next() Row
 
-	// End returns the invalid end Row.
-	End() Row
+	// HasNext returns true if the iteration has more elements.
+	HasNext() bool
```

Comment:
How about that we use `Valid` instead of `HasNext`, then add an `Error` method to check if there is any error.
Then we don't need `NextStrict` and `HasNextStrict`.